### ISSUE TITLE
onchange js event fired in WebElement.clear()

### ISSUE THREAD

**Issue by slaout** [SeleniumHQ/selenium#1841]
## Meta -

OS: Windows 7
Selenium Version: 2.53.0
Browser: Firefox
Browser Version: 45 (64-bit)
## Expected Behavior -

When calling webElement.clear(), I expect the following JavaScript events to be fired:
onfocus (only if the field did not have focus before)
oncut (ideally, to simulate what a real user would do to clear, but I don't care that much about this event)
oninput (and not onchange, because focus is still on the field)
## Actual Behavior -

onfocus (only if the field did not have focus before)
onchange (instead of oninput)
## Steps to reproduce -

``` html
<input id="test" oninput="console.log('input')" onchange="console.log('change')">
```

``` java
driver.findElement(By.id("test")).clear();
```

In the Firefox console log, you see "change" instead of "input".

**Comment by slaout** [comment]
As a temporary workaround, I added this method (gathered on the Web, and enhanced to support textarea too):

```java
    /**
     * Replaces {@link WebElement#clear()} in order to simulate what a real user would do to clear an input text field, by not calling onchange
     * JavaScript event, and by calling oninput and keystroke events (Ctrl+Home, Ctrl+Shift+End, BackSpace).
     * <p>
     * {@link WebElement#clear()} is buggy because it fires onfocus (if not already focused) and then onchange! The onchange event must be fired
     * only after the input value has changed AND it has lost focus. When clearing, the input do not lose focus, so Selenium is buggy here. It
     * must trigger oninput instead (and it doesn't). Here is the initial issue closed by mistake without fix:
     * https://github.com/SeleniumHQ/selenium-google-code-issue-archive/issues/214 and here is the new issue I created to solve that for good:
     * https://github.com/SeleniumHQ/selenium/issues/1841
     * 
     * @param element a text INPUT or TEXTAREA field
     * @return this SeleniumActions for method chaining
     */
    public SeleniumActions clear(final WebElement element) {
        final Actions navigator = new Actions(seleniumDriver);
        navigator.click(element) //
                .keyDown(Keys.CONTROL) //
                /* */.sendKeys(Keys.HOME) //
                /* */.keyDown(Keys.SHIFT) //
                /* * * */.sendKeys(Keys.END) //
                /* */.keyUp(Keys.SHIFT) //
                .keyUp(Keys.CONTROL) //
                .sendKeys(Keys.BACK_SPACE).perform();
        return this;
    }
```

In the wait of a fix, I replaced all my WebElement.clear() methods with this homemade one, because I can't trust Selenium's one.

**Comment by s2t2** [comment]
That's weird, I'm actually expecting `clear()` to trigger `onChange`, but it's not happening. Wish we could switch places!

**Comment by drjayvee** [comment]
I'm running into issues modernizing my code to use the [`input` event](https://developer.mozilla.org/en-US/docs/Web/Events/input). I expect the `clear` action to cause the `input` event to fire, which is not the case.

The [WebDriver spec for the `clear` action](https://w3c.github.io/webdriver/webdriver-spec.html#element-clear) actually says "2. Fire change event", so I'm afraid this will not get "fixed", as its a problem with the spec, not the implementation.

I would argue that `change` should _not_ fire, but `input` _should_.

**Comment by lukeis** [comment]
i would recommend someone logging a w3c spec bug too :)

https://github.com/w3c/webdriver/issues

**Comment by KillaPL** [comment]
I'm gitting this issue when testing react app through ruby wrapper. When setting input to empty string, onChange events are not fired. What are to possible ways here? I've written a workaround for me too, but this clearly looks like this should happen on seleniums side.

**Comment by drjayvee** [comment]
I filed an issue with the spec: [Element clear command should fire input event instead of change](https://github.com/w3c/webdriver/issues/445)

**Comment by TylerNielsen** [comment]
I'm also running into this issue. I just updated WebDriver from 3.7 to 3.14 and many of our tests are now failing. In our "SetValue" method, we call element.Clear() before calling element.SendKeys(). However, now when clear is called, eventhandlers are being executed that interpret an empty input field as "0.00". Subsequent call to element.SendKeys does not input the new value in the field. Instead "0.00" stays.

Workaround until this is fixed is similar to slaout's suggestion, but I prefer to not Action class.


```
public void ClearElement(IWebElement element)
{
	// Our method includes waits and handling to ensure element is not stale, but excluded
        // for simplicity.
        element.SendKeys(Keys.Control + "a");
	element.SendKeys(Keys.Delete);
}

```

**Comment by diemol** [comment]
I went and read all the linked issues to understand why this issue has staled, and it is clear that the spec says one [thing](https://www.w3.org/TR/webdriver1/#element-clear) and this [comment](https://github.com/w3c/webdriver/issues/445#issuecomment-273187301) ratifies it.
Based on the previous, I will close this issue, and any alternate behaviour should be raised to the spec first.
